feat!: split apart checkbox checked and indeterminate props

ahuth opened the pull request:
EFI-949

### Summary:

Breaking change! (although I'm not aware of any consumers affected by it)

- Give the Checkbox component separate `checked` and `indeterminate` props.
- Also makes `indeterminate` checkboxes officially indeterminate in the DOM.

### Why:

Thought about this while working on #1511.

https://html.spec.whatwg.org/multipage/input.html#checkbox-state-(type=checkbox) calls out `indeterminate` as a separate piece of state than `checked`.

> The control is never a true tri-state control, even if the element's [indeterminate](https://html.spec.whatwg.org/multipage/input.html#dom-input-indeterminate) IDL attribute is set to true. The [indeterminate](https://html.spec.whatwg.org/multipage/input.html#dom-input-indeterminate) IDL attribute only gives the appearance of a third state.

Practically this means a checkbox can be in all 4 combinations of `checked` and `indeterminate` (although there are only 3 visual states, because both indeterminates look the same).

Verified this is how browsers behave with https://codepen.io/ahuth/pen/MWqmwyv.

### Test Plan:

<!--
  How did you validate that your changes were implemented correctly?
-->

- [x] Wrote [automated tests](https://czi.atlassian.net/wiki/x/Hbl1H)
- [ ] CI tests / new tests are not applicable
- [ ] Manually tested my changes, but I want to keep the details secret
- [x] Manually tested my changes, and here are the details:
  - Storybook

ahuth commented on src/components/Checkbox/__snapshots__/Checkbox.test.tsx.snap:
```diff
@@ -137,8 +137,6 @@ exports[`<Checkbox /> Disabled story renders snapshot 1`] = `
             class="checkbox"
           >
             <input
-              aria-checked="mixed"
```
Don't need `aria-checked` when we set `indeterminate` on the DOM node.

ahuth commented on src/components/CheckboxInput/CheckboxInput.tsx:
```diff
@@ -31,32 +35,24 @@ export type CheckboxInputProps = CheckboxHTMLElementProps & {
-  const checkedProps =
-    checked === 'indeterminate'
-      ? {
-          'aria-checked': 'mixed' as const,
```
Previously we set `aria-checked` so screen readers knew about it, but the DOM itself didn't consider the checkbox to be indeterminate.

Here we fix that (which unfortunately requires JS and jumping through a hoop).

ahuth commented on src/components/CheckboxInput/CheckboxInput.tsx:
```diff
@@ -31,32 +35,24 @@ export type CheckboxInputProps = CheckboxHTMLElementProps & {
-    checked === 'indeterminate'
-      ? {
-          'aria-checked': 'mixed' as const,
-          checked: true,
```
Previously we considered indeterminate checkboxes as "checked".

That's not correct, according to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate_state_checkboxes.

Indeterminate checkboxes should be considered un-checked.

jinlee93 replied:
Think `'indeterminate'` should be removed from the argtypes `checked` options

ahuth commented on src/components/CheckboxInput/CheckboxInput.tsx:
```diff
@@ -31,32 +35,24 @@ export type CheckboxInputProps = CheckboxHTMLElementProps & {
+>(({ checked, className, disabled, indeterminate, ...other }, ref) => {
+  const forwardedRef = useForwardedRef(ref);
+
+  // Make this checkbox indeterminate. Can only be done with JS for some reason.
```
See https://github.com/facebook/react/issues/1798 for why React hasn't added an `indeterminate` prop to input checkboxes

ahuth commented:
How code coverage decrease when I added tests? 🤔

jinlee93 commented:
> How test coverage decrease when I added tests? 🤔

I think coverage is calculated by Hits / Lines and your new tests covered 11 / 12 lines (91.66%) which technically brought the average down LOL

ahuth commented:
@chanzuckerberg/edu-frontend-infra

booc0mtaco commented:
@ahuth can you update the PR title to be semver? looks like github made the checkbox PR commit [non-compliant](https://github.com/chanzuckerberg/edu-design-system/commit/6fc50f53b02301789af2aa578e27e31ec970bc06) on merge 

It's annoying that it defaults to the title if there are multiple commits on a PR branch...

ahuth commented:
Oh, the PR title needs to adhere to conventional-commits? Guess I assumed that the defaults for merges were fine. Maybe that's why we used to only allow rebase merges (since commit messages are enforced).

booc0mtaco commented:
I might be missing what thing(s) break with this change? is it the place where we try and handle `checked="indeterminate"`? Most of it looks kinda additive to my eye

booc0mtaco commented:
> Oh, the PR title needs to adhere to conventional-commits? Of course!
> 
> Guess I assumed that the defaults for merges were fine. Maybe that's why we used to only allow rebase merges (since commit messages are enforced) 🤔

Yah, probably. It's a REALLY annoying "feature" in GitHub where it does the Right Thing (tm) if the branch has one commit to squash and merge, but this behavior when it has multiple ones. I'd honestly prefer if it left the title field blank when ambiguous but 🤷. That code is good, so we can update changelog manually if we want; no biggie

ahuth commented:
Updated the title.

> what thing(s) break with this change?

The `checked` prop changed from `true | false | `indeterminate` to just `true | false`. (And added an indeterminate boolean, but that's not breaking).

Any usages using `checked="indeterminate"` would need to be updated... although I don't believe there are any in the wild.

booc0mtaco commented on src/components/CheckboxInput/CheckboxInput.tsx:
```diff
@@ -9,11 +10,9 @@ type CheckboxHTMLElementProps = Omit<
+   * Whether checkbox is checked.
    */
-  checked?: boolean | 'indeterminate';
+  checked?: boolean;
```
The breaking change => reducing the available options

booc0mtaco reviewed (approved):
Thanks for patching this up! I think this PR will have enough details for anyone to sort out what changes they should make in the off chance they were trying to use indeterminate directly.

jinlee93 commented on src/components/Checkbox/Checkbox.stories.tsx:
```diff
@@ -115,10 +122,10 @@ export const WithoutVisibleLabel: StoryObj<Args> = {
       <Checkbox {...args} disabled />
       <Checkbox {...args} checked disabled />
-      <Checkbox {...args} checked="indeterminate" disabled />
+      <Checkbox {...args} disabled indeterminate />
```
I like how simple this looks

jinlee93 commented on src/components/CheckboxInput/CheckboxInput.test.tsx:
```diff
@@ -0,0 +1,29 @@
+    render(<CheckboxInput id="test" />);
+    const checkbox = screen.getByRole<HTMLInputElement>('checkbox');
+    expect(checkbox.indeterminate).toEqual(false);
+  });
```
Just also manually tested <Checkbox indeterminate> and then checking it to true and seeing that the input.indeterminate === false

ahuth replied:
Sweet, thanks!

jinlee93 commented on src/util/useForwardedRef.ts:
```diff
@@ -0,0 +1,21 @@
+import { useRef, useEffect, type ForwardedRef, type RefObject } from 'react';
+
+/**
+ * Safely access a forwarded ref, which may or may not be present.
+ */
```
nice

ahuth commented:
Thanks!
